py/server/tests/test_table_listener.py

```diff
@@ -3,101 +3,162 @@
 #
 import time
 import unittest
+from pprint import pprint
+from types import SimpleNamespace
+from typing import List, Dict, Union
+
+import numpy
+from deephaven.experimental import time_window
 
 from deephaven import time_table
+from deephaven.jcompat import to_sequence
 from deephaven.table_listener import listen, TableListener
-
 from tests.testbase import BaseTestCase
 
 
 class TableListenerTestCase(BaseTestCase):
+
     def setUp(self) -> None:
         with self.ugp_lock_exclusive():
-            self.table = time_table("00:00:01").update(["X=i"]).tail(5)
-        self.table_change_processed = None
-
-    def test_shift_oblivious_listener_obj(self):
+            self.test_table = time_table("00:00:00.001").update(["X=i%2"]).sort("X").tail(16)
+            source_table = time_table("00:00:00.001").update(["TS=currentTime()"])
+            self.test_table2 = time_window(source_table, ts_col="TS", window=10 ** 7, bool_col="InWindow")
+
+    def tearDown(self) -> None:
+        self.test_table = None
+        self.test_table2 = None
+
+    def check_result(self, changes: List[Dict[str, numpy.ndarray]], cols: Union[str, List[str]] = None):
+        if not changes:
+            return
+        cols = to_sequence(cols)
+        for change in changes:
+            self.assertTrue(isinstance(change, dict))
+            if not cols:
+                cols = [col.name for col in self.test_table.columns]
+            for col in cols:
+                self.assertIn(col, change.keys())
+                self.assertTrue(isinstance(change[col], numpy.ndarray))
+                self.assertEqual(change[col].ndim, 1)
+                # print(change[col])
+
+    def test_listener_obj(self):
         class ListenerClass(TableListener):
             def __init__(self):
-                self.table_change_added = None
-                self.table_change_removed = None
-                self.table_change_modified = None
-
-            def onUpdate(self, added, removed, modified):
-                self.table_change_added = added
-                self.table_change_removed = removed
-                self.table_change_modified = modified
-
-            def has_table_changed(self):
-                return (
-                    self.table_change_removed
-                    or self.table_change_modified
-                    or self.table_change_added
-                )
+                self.table_update = SimpleNamespace(added=None, removed=None, modified=None, modified_prev=None,
+                                                    Shifted=None)
+
+            def onUpdate(self, is_replay, update):
+                update.chunk_size = 4
```

Comment:
I see here that this is how you are intending to set the `chunk_size`.  If you are going this route, do you need to be carful that the chunk size isn't changed after data has started to be pulled?

Reply:
I'm also conflicted on this kind of API.  Ideally `update` is some sort of immutable that simply contains what has changed.  We are not that clean, because we are caching in the object.  Is there a way to get an API that maintains logical immutability while still providing a concise and clean user API?  The answer may be no, and we might have to accept some logical contamination to make it easy to use.

Reply:
I guess these API questions also come back to the "give me everything as a single array" question.  It raises other questions about if we always want the chunk sizes to be the same for different columns.